File: gdscript/src/main/java/gdscript/psi/impl/GdPatternImpl.java
package gdscript.psi.impl;

import com.intellij.extapi.psi.ASTWrapperPsiElement;
import com.intellij.lang.ASTNode;
import com.intellij.psi.PsiElementVisitor;
import com.intellij.psi.util.PsiTreeUtil;
import gdscript.psi.GdArrayPattern;
import gdscript.psi.GdBindingPattern;
import gdscript.psi.GdDictPattern;
import gdscript.psi.GdExpr;
import gdscript.psi.GdPattern;
import gdscript.psi.GdVisitor;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class GdPatternImpl extends ASTWrapperPsiElement implements GdPattern {

  public GdPatternImpl(@NotNull ASTNode node) {
    super(node);
  }

  public void accept(@NotNull GdVisitor visitor) {
    visitor.visitPattern(this);
  }

  @Override
  public void accept(@NotNull PsiElementVisitor visitor) {
    if (visitor instanceof GdVisitor) accept((GdVisitor)visitor);
    else super.accept(visitor);
  }

  @Override
  @Nullable
  public GdArrayPattern getArrayPattern() {
    return PsiTreeUtil.getChildOfType(this, GdArrayPattern.class);
  }

  @Override
  @Nullable
  public GdBindingPattern getBindingPattern() {
    return PsiTreeUtil.getChildOfType(this, GdBindingPattern.class);
  }

  @Override
  @Nullable
  public GdDictPattern getDictPattern() {
    return PsiTreeUtil.getChildOfType(this, GdDictPattern.class);
  }

  @Override
  @Nullable
  public GdExpr getExpr() {
    return PsiTreeUtil.getChildOfType(this, GdExpr.class);
  }

}

File: gdscript/src/main/kotlin/gdscript/psi/impl/GdPatternImpl.kt
package gdscript.psi.impl

import com.intellij.extapi.psi.ASTWrapperPsiElement
import com.intellij.lang.ASTNode
import com.intellij.psi.PsiElementVisitor
import com.intellij.psi.util.PsiTreeUtil
import gdscript.psi.GdArrayPattern
import gdscript.psi.GdBindingPattern
import gdscript.psi.GdDictPattern
import gdscript.psi.GdExpr
import gdscript.psi.GdPattern
import gdscript.psi.GdVisitor

class GdPatternImpl(node: ASTNode) : ASTWrapperPsiElement(node), GdPattern {
    fun accept(visitor: GdVisitor) {
        visitor.visitPattern(this)
    }

    override fun accept(visitor: PsiElementVisitor) {
        if (visitor is GdVisitor) accept(visitor)
        else super.accept(visitor)
    }

    override val arrayPattern: GdArrayPattern?
        get() = PsiTreeUtil.getChildOfType(this, GdArrayPattern::class.java)

    override val bindingPattern: GdBindingPattern?
        get() = PsiTreeUtil.getChildOfType(this, GdBindingPattern::class.java)

    override val dictPattern: GdDictPattern?
        get() = PsiTreeUtil.getChildOfType(this, GdDictPattern::class.java)

    override val expr: GdExpr?
        get() = PsiTreeUtil.getChildOfType(this, GdExpr::class.java)
}
